### PR TITLE
Grid JS to TS conversion

### LINKED ISSUE #21: Convert all js in grid to ts
Convert everything in the grid package to TypeScript. This will be particularly useful with the grid metrics and just helping grid usage.

### PR DESCRIPTION
**Pull request by mofojed**
- Converted the rest of @deephaven/grid package to TypeScript
- Separated GridModel into an abstract class, and two separate interfaces, `ExpandableGridModel`, `EditableGridModel`
- Fixes #21

### CODE REVIEW

File: packages/grid/src/GridRenderer.ts [hunk 1]
```diff
@@ -1710,7 +1930,20 @@ class GridRenderer {
...
+    state: GridRenderState,
+    viewport: {
+      left?: VisibleIndex;
+      top?: VisibleIndex;
+      right?: VisibleIndex;
+      bottom?: VisibleIndex;
+      minX?: VisibleIndex;
+      minY?: VisibleIndex;
+      maxX?: VisibleIndex;
+      maxY?: VisibleIndex;
```
Comment (mattrunyon) [inline packages/grid/src/GridRenderer.ts:1944]: Are these actually indexes? Looks like the default values set are coordinates/pixels

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/grid/src/EditableGridModel.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
+import GridRange from './GridRange';
+import { ModelIndex } from './GridMetrics';
+import GridModel from './GridModel';
+
+export function isEditableGridModel(
+  model: GridModel
+): model is EditableGridModel {
+  return (model as EditableGridModel)?.isEditable !== undefined;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/EditableGridModel.ts:8]: Eh I'm just going by the Type guarding example: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards
I don't think it makes much of a difference other than being more specific, but I'll update

File: packages/grid/src/EditableGridModel.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+   * @returns True if it's a valid value, false otherwise
+   */
+  isValidForCell(column: ModelIndex, row: ModelIndex, value: string): boolean;
+
+  /**
+   * Delete ranges from an editable grid
+   * @param ranges The ranges to delete
+   * @returns A promise that resolves successfully when the operation is complete or rejects if there's an error
+   */
+  delete(ranges: GridRange[]): Promise<void>;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/EditableGridModel.ts:93]: Yes

File: packages/grid/src/KeyHandler.ts [hunk 4]
```diff
@@ -6,16 +6,26 @@
...
+export type KeyHandlerResultOptions = {
+  stopPropagation?: boolean;
+  preventDefault?: boolean;
+};
+
+/**
+ * Result from the key handler for an event.
+ * Return `false` to pass the event on, otherwise the event is consumed.
+ */
+export type KeyHandlerResult = boolean | KeyHandlerResultOptions;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/KeyHandler.ts:25]: I had a separate type because in Grid need to cast it to that other object type with `stopPropagation` and `preventDefault`, seemed cleaner. 
There are some cases we return either false or true, or the object.
I've updated the comment but I'm not sure updating the name would make much of a difference?

File: packages/grid/src/GridModel.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    return true;
+  }
+
+  /**
+   * @param row Row to check
+   * @returns True if the row is movable
+   */
+  isRowMovable(row: ModelIndex): boolean {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridModel.ts:128]: The column/row moves don't affect the underlying model at all, so it should work for any model... figure by having it defaulted to true the functionality is more discoverable

File: packages/grid/src/GridMouseHandler.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  stopPropagation?: boolean;
+  preventDefault?: boolean;
+};
+/**
+ * Result from the mouse handler for an event.
+ * Return `false` to pass the event on, otherwise the event is consumed.
+ */
+export type GridMouseHandlerResult = boolean | GridMouseHandlerResultOptions;
+
+export type GridMouseHandlerFunctionName =
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridMouseHandler.ts:21]: Yea it's just one mouse handler can map to all of these functions, it's not just one type.

File: packages/grid/src/GridMouseHandler.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  | 'onContextMenu'
+  | 'onDoubleClick'
+  | 'onUp'
+  | 'onWheel';
+
+/**
+ * Define a region in the grid that interacts with the mouse on a grid.
+ * Return true from any of the events to indicate they're consumed, and stopPropagation/preventDefault will be called.
+ */
+export class GridMouseHandler {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridMouseHandler.ts:36]: I don't think so. An event handler would only handle one event (onDown, or onClick, etc). A GridMouseHandler is one class that has all the logic for an interaction, and can hook into one or more events.

File: packages/grid/src/GridUtils.test.ts [hunk 8]
```diff
@@ -155,14 +168,14 @@ describe('floating item checks', () => {
...
-    movedItems = [],
+    start: GridRangeIndex,
+    end: GridRangeIndex,
+    movedItems: MoveOperation[] = [],
     expectedResult = [[start, end]]
   ) {
     expect(
       GridUtils.getModelRangeIndexes(start, end, movedItems).sort(
-        (a, b) => a[0] - b[0]
+        (a, b) => (a as AxisRange)[0] - (b as AxisRange)[0]
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridUtils.test.ts:178]: The function returns a `GridAxisRange`, which could have `null` values. This casts it to `AxisRange` so no `null` values

File: packages/grid/src/GridUtils.test.ts [hunk 9]
```diff
@@ -217,23 +230,23 @@ describe('start/end range adjustment in one dimension', () => {
...
+    movedRows: MoveOperation[] = [],
     expectedRanges = ranges
   ) {
     expect(
       GridUtils.getModelRanges(ranges, movedColumns, movedRows).sort((a, b) =>
         a.startColumn !== b.startColumn
-          ? a.startColumn - b.startColumn
-          : a.startRow - b.endRow
+          ? (a.startColumn as ModelIndex) - (b.startColumn as ModelIndex)
+          : (a.startRow as ModelIndex) - (b.endRow as ModelIndex)
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridUtils.test.ts:242]: `GridRangeIndex` could be `null`, this is saying they won't be `null`

File: packages/grid/src/mouse-handlers/EditMouseHandler.ts [hunk 10]
```diff
@@ -1,18 +1,22 @@
...
 
 /**
  * Handles clicking on a cell to edit it in an editable grid
  */
 class EditMouseHandler extends GridMouseHandler {
-  onDoubleClick(
-    { column, row }: { column: number; row: number },
-    grid: Grid
-  ): boolean {
+  onDoubleClick({ column, row }: GridPoint, grid: Grid): boolean {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/EditMouseHandler.ts:12]: Yea, boolean is just a subset of that which is all this function returns, I should match the parent though I think

File: packages/grid/src/mouse-handlers/GridSeparatorMouseHandler.ts [hunk 11]
```diff
@@ -1,86 +1,104 @@
...
+  abstract firstIndexProperty: FirstIndexProperty;
 
-  firstIndexProperty = OVERRIDE_STRING;
+  abstract treePaddingProperty: TreePaddingProperty;
 
-  treePaddingProperty = OVERRIDE_STRING;
+  abstract getHiddenItems(
+    itemIndex: VisibleIndex,
+    metrics: GridMetrics
+  ): VisibleIndex[];
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridSeparatorMouseHandler.ts:67]: Yes

File: packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts [hunk 12]
```diff
@@ -1,23 +1,32 @@
...
 class GridSelectionMouseHandler extends GridMouseHandler {
-  startPoint = null;
+  private startPoint?: GridPoint;
 
-  hasExtendedFloating = false;
+  private hasExtendedFloating = false;
 
   // Timer used when holding a drag past the end of the grid
-  timer = null;
+  private timer?: ReturnType<typeof setTimeout>;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts:15]: Yea we did this elsewhere so just matching

File: packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts [hunk 13]
```diff
@@ -187,21 +207,31 @@ class GridSelectionMouseHandler extends GridMouseHandler {
...
+    deltaY: number
+  ): void {
     const { row, column } = gridPoint;
     const { metrics } = grid;
+    if (!metrics) throw new Error('metrics not set');
+
     const { selectionEndRow, selectionEndColumn } = grid.state;
+    if (selectionEndRow == null || selectionEndColumn == null)
+      throw new Error('selection not set');
+
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts:234]: Yea with Prettier formatting I think it makes it a lot clearer - if you try and have multiple things here, it'll bump it down. I'll add Curly here but we can enable option in the separate commit (maybe just the `multi` option)

File: packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts [hunk 14]
```diff
@@ -229,9 +264,11 @@ class GridSelectionMouseHandler extends GridMouseHandler {
...
   }
 
-  stopTimer() {
-    clearTimeout(this.timer);
-    this.timer = null;
+  stopTimer(): void {
+    if (this.timer !== undefined) {
+      clearTimeout(this.timer);
+      this.timer = undefined;
+    }
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts:271]: window.clearTimeout is expecting a number, so would need to do window.setTimeout as well above. I kind of like having the typing be a Timeout type vs just a number?

File: packages/grid/src/mouse-handlers/GridRowSeparatorMouseHandler.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+  marginProperty: MarginProperty = 'columnHeaderHeight';
+
+  calculatedSizesProperty: CalculatedSizeProperty = 'calculatedRowHeights';
+
+  modelIndexesProperty: ModelIndexesProperty = 'modelRows';
+
+  firstIndexProperty: FirstIndexProperty = 'firstRow';
+
+  treePaddingProperty: TreePaddingProperty = 'treePaddingY';
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridRowSeparatorMouseHandler.ts:56]: Cool!

File: packages/grid/src/mouse-handlers/GridRowSeparatorMouseHandler.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    metricCalculator: GridMetricCalculator,
+    modelIndex: ModelIndex
+  ): void {
+    metricCalculator.resetRowHeight(modelIndex);
+  }
+
+  updateSeparator(grid: Grid, separatorIndex: VisibleIndex | null): void {
+    grid.setState({
+      draggingRowSeparator: separatorIndex,
+      isDragging: !!separatorIndex,
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridRowSeparatorMouseHandler.ts:80]: It is not, but I'm going to switch to !== null anyway since that seems clearer

File: packages/grid/src/mouse-handlers/GridColumnSeparatorMouseHandler.ts [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    metricCalculator: GridMetricCalculator,
+    modelIndex: ModelIndex
+  ): void {
+    metricCalculator.resetColumnWidth(modelIndex);
+  }
+
+  updateSeparator(grid: Grid, separatorIndex: VisibleIndex | null): void {
+    grid.setState({
+      draggingColumnSeparator: separatorIndex,
+      isDragging: !!separatorIndex,
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/mouse-handlers/GridColumnSeparatorMouseHandler.ts:80]: No but updating anyway

File: packages/grid/src/Grid.tsx [hunk 18]
```diff
@@ -57,23 +216,73 @@ class Grid extends PureComponent {
...
+  /**
+   * On some devices there may be different scaling required for high DPI. Get the scale required for the canvas.
+   * @param context The canvas context
+   * @returns The scale to use
+   */
+  static getScale(context: CanvasRenderingContext2D): number {
     const devicePixelRatio = window.devicePixelRatio || 1;
+
+    // backingStorePixelRatio is deprecated, but check it just in case
+    const oldContext = context as OldCanvasRenderingContext2D;
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:228]: Yes I like

File: packages/grid/src/Grid.tsx [hunk 19]
```diff
@@ -308,41 +538,61 @@ class Grid extends PureComponent {
...
-      passive: false,
-    });
+    this.canvas?.removeEventListener(
+      'wheel',
+      (this.handleWheel as unknown) as EventListenerOrEventListenerObject,
+      // Need to cast this because passive is not in EventListenerOptions for some reason
+      ({
+        passive: false,
+      } as unknown) as EventListenerOptions
+    );
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:553]: Yea I don't think I need to provide the passive option here for removing, good call.

File: packages/grid/src/Grid.tsx [hunk 20]
```diff
@@ -308,41 +538,61 @@ class Grid extends PureComponent {
...
 
-  getTheme() {
+  getTheme(): GridThemeType {
     const { theme } = this.props;
     return Grid.getTheme(theme);
   }
 
-  getGridPointFromEvent(event) {
+  getGridPointFromEvent(event: MouseEvent): GridPoint {
+    if (!this.canvas) throw new Error('Canvas must be set');
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:576]: Yea true

File: packages/grid/src/Grid.tsx [hunk 21]
```diff
@@ -929,25 +1250,40 @@ class Grid extends PureComponent {
...
-   * @param {string[][] | string} value Table or a string that is being pasted
+   * @param value Table or a string that is being pasted
    */
-  async pasteValue(value) {
+  async pasteValue(value: string[][] | string): Promise<void> {
     const { model } = this.props;
     const { movedColumns, movedRows, selectedRanges } = this.state;
 
     try {
+      if (!isEditableGridModel(model)) throw new Error('model is not editable');
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1286]: Ouuuu or type assertions! I'm going to do that instead.

File: packages/grid/src/Grid.tsx [hunk 22]
```diff
@@ -1027,21 +1363,39 @@ class Grid extends PureComponent {
...
   }
 
-  setValueForCell(column, row, value) {
+  /**
+   * Set a value to a specific cell. If the value is not valid for that cell, do not set it
+   * @param column Column index to set the value for
+   * @param row Row index to set the value for
+   * @param value Value to set at that cell
+   */
+  setValueForCell(
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1376]: Yea I'll add boolean return value

File: packages/grid/src/Grid.tsx [hunk 23]
```diff
@@ -1208,32 +1598,37 @@ class Grid extends PureComponent {
...
     for (let i = 0; i < keyHandlers.length; i += 1) {
       const keyHandler = keyHandlers[i];
-      const result = keyHandler.onDown(e, this);
+      const result = keyHandler.onDown(event, this);
       if (result) {
-        if (result?.stopPropagation ?? true) e.stopPropagation();
-        if (result?.preventDefault ?? true) e.preventDefault();
+        const options = result as KeyHandlerResultOptions;
+        if (options?.stopPropagation ?? true) event.stopPropagation();
+        if (options?.preventDefault ?? true) event.preventDefault();
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1613]: Why would it break here? `{ stopPropagation: false, preventDefault: false}` is still legit if you want to consume it so no other registered grid mouse handlers can use it, but you want the event to bubble up.

undefined will default to true.

I don't think any was is particularly clearer than the other so I'm going to leave it

File: packages/grid/src/Grid.tsx [hunk 24]
```diff
@@ -1251,8 +1646,9 @@ class Grid extends PureComponent {
         }
 
         // result is bool or object, events are stopped by default
-        if (result?.stopPropagation ?? true) event.stopPropagation();
-        if (result?.preventDefault ?? true) event.preventDefault();
+        const options = result as GridMouseHandlerResultOptions;
+        if (options?.stopPropagation ?? true) event.stopPropagation();
+        if (options?.preventDefault ?? true) event.preventDefault();
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1651]: Same as above

File: packages/grid/src/Grid.tsx [hunk 25]
```diff
@@ -1468,26 +1870,44 @@ class Grid extends PureComponent {
...
 
-  handleEditCellChange(value) {
+  /**
+   * Handle a change in the value in an editing cell
+   * @param value New value set
+   */
+  handleEditCellChange(value: string): void {
     this.setState(({ editingCell }) => ({
-      editingCell: { ...editingCell, value },
+      editingCell: { ...editingCell, value } as EditingCell,
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1891]: Added a type assertion - this function should never be able to be called if editingCell is null.

File: packages/grid/src/GridRenderer.ts [hunk 26]
```diff
@@ -511,30 +594,39 @@ class GridRenderer {
           ? visibleColumnXs.get(floatingLeftColumnCount + 1)
           : -10,
       minY:
-        floatingTopColumnCount > 0 &&
-        visibleRowYs.has(floatingTopColumnCount + 1)
-          ? visibleRowYs.get(floatingTopColumnCount + 1)
+        floatingTopRowCount > 0 && visibleRowYs.has(floatingTopRowCount + 1)
+          ? visibleRowYs.get(floatingTopRowCount + 1)
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridRenderer.ts:598]: Because those ranges may not actually be visible, they may be off screen

[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridRenderer.ts:598]: Errrrr wait. I'm already checking for `.has` so the getOrThrow will never throw anyway - I think it's because if it's undefined then the addition operation in the below would give an error, so I wasn't getting an error here. I'll add it just to be consistent

File: packages/grid/src/GridRenderer.ts (hunk 1 above)
Comment (mofojed) [reply packages/grid/src/GridRenderer.ts:1944]: They are indeed `Coordinate`, nice catch.

File: packages/grid/src/EditableGridModel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/EditableGridModel.ts:8]: This is fine then. Didn't realize it was the example TS provided

File: packages/grid/src/Grid.tsx [hunk 27]
```diff
@@ -889,7 +1208,7 @@ class Grid extends PureComponent {
       this.isStuckToRight = isStickyRight && left >= lastLeft;
     }
 
-    this.setState(viewState);
+    this.setState(viewState as GridState);
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1211]: Nevermind, in some places I call this with other state vars to update. Just going to leave Partial for now and can address later if needed.

File: packages/grid/src/EditableGridModel.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/EditableGridModel.ts:104]: Can you clarify which behavior in the comment?

I'm assuming it nulls out the values in the selection and not the Excel "delete range" which then asks if you want to shift data left/up/etc.

File: packages/grid/src/GridMouseHandler.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/GridMouseHandler.ts:23]: Gotcha

File: packages/grid/src/GridUtils.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/GridUtils.test.ts:178]: The cast doesn't actually mean there's no `null` values though, so wouldn't it be possible that you end up w/ `null` values since the function returns a possible `null`?

File: packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts]: Sounds good

[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts]: Made #353 to track

File: packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/mouse-handlers/GridSelectionMouseHandler.ts:273]: Ya it's more robust this way

File: packages/grid/src/EditableGridModel.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/EditableGridModel.ts:104]: True. And in fact, I don't even have this wired up in @deephaven/grid at all, so I'm just going to move the delete function into `IrisGridModel`... until we have some way from Grid itself to do this delete row function, it doesn't make sense to have in grid I don't think. And then it can take a param to ask if it should shift.
And I'm not going to change the signature or typescriptify IrisGridModel in this change either - will decide to clean that up at this time.

To null/blank out the values in a selection we already do that by calling `setValueForRanges`.

Anyway, good catch!

File: packages/grid/src/Grid.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/Grid.tsx]: Is `{ stopPropagation: false, preventDefault: false }` different than just returning `false` then? I would have expected the 2 to be equivalent. I guess it just wasn't that clear that returning `{ stopPropagation: false, preventDefault: false }` from the handler will still consume the event as far as the grid mouse handler is consumed. Shouldn't they at least `stopPropagation` in that case?

File: packages/grid/src/GridUtils.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridUtils.test.ts:178]: Well, I guess, in which case the test would fail, since we're not expecting null here... do I need to rewrite it to explicitly check for null, or is casting with what's expected of the test knowing that it'll fail the test if it's null is good enough?

File: packages/grid/src/Grid.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/Grid.tsx]: Yes, returning the object still stops it from being passed to other registered grid mouse handlers, but could still allow the mouse event to propagate/bubble up through the regular event system.

Adding some more comments to EventHandlerResult to try and make all the options clear

File: packages/grid/src/Grid.tsx (hunk 27 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/Grid.tsx:1197]: Ahh I guess I didn't see the error where it was called when I tried it locally

File: packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts [hunk 28]
```diff
@@ -65,11 +65,11 @@ class MockIrisGridTreeModel
     return this.isEditable;
   }
 
-  setValues(edits: EditOperation[]): Promise<void> {
+  setValues(edits: EditOperation[]): Promise<never> {
     throw new Error('Method not implemented.');
   }
 
-  delete(ranges: GridRange[]): Promise<void> {
+  delete(ranges: GridRange[]): Promise<never> {
```
Comment (mattrunyon) [inline packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts:72]: I think these can just be `never` and it should be fine? It would just prevent you from trying to do anything w/ these operations when using the mock I believe

Looks like we would also want to add a flag to TS (or eslint) to make this an error for unreachable code. When I tried it out just now as unreachable, I don't get an error in TS. Looks like TS may allow it by default, but eslint disabled the rule since it used to be the default tsc behavior was to error on unreachable code.

https://github.com/typescript-eslint/typescript-eslint/issues/1041

File: packages/grid/src/Grid.test.tsx [hunk 29]
```diff
@@ -109,14 +109,14 @@ function getClientY(
 function mouseEvent(
   column: VisibleIndex,
   row: VisibleIndex,
-  fn: (event: MouseEvent) => void,
+  fn: (event: React.MouseEvent) => void,
```
Comment (mattrunyon) [inline packages/grid/src/Grid.test.tsx:112]: Can just be `React.MouseEventHandler` I believe

File: packages/grid/src/Grid.test.tsx [hunk 30]
```diff
@@ -149,7 +149,7 @@ function mouseMove(
   mouseEvent(
     column,
     row,
-    component.handleMouseDrag,
+    (component.handleMouseDrag as unknown) as (event: React.MouseEvent) => void,
```
Comment (mattrunyon) [inline packages/grid/src/Grid.test.tsx:152]: `React.MouseEventHandler`

File: packages/grid/src/Grid.test.tsx [hunk 31]
```diff
@@ -168,7 +168,7 @@ function mouseUp(
   mouseEvent(
     column,
     row,
-    component.handleMouseUp,
+    (component.handleMouseUp as unknown) as (event: React.MouseEvent) => void,
```
Comment (mattrunyon) [inline packages/grid/src/Grid.test.tsx:171]: `React.MouseEventHandler`

File: packages/grid/src/Grid.tsx [hunk 32]
```diff
@@ -1682,43 +1662,44 @@ class Grid extends PureComponent<GridProps, GridState> {
...
     this.notifyMouseHandlers('onLeave', event, false);
     this.setState({ mouseX: null, mouseY: null });
   }
 
   handleMouseDrag(event: MouseEvent): void {
     this.setState({ isDragging: true });
-    this.notifyMouseHandlers('onDrag', event, true, true);
+    this.notifyMouseHandlers(
+      'onDrag',
+      (event as unknown) as React.MouseEvent,
```
Comment (mattrunyon) [inline packages/grid/src/Grid.tsx:1682]: Can the event param just be a `React.MouseEvent`? Looks like you have that set for some other params here

File: packages/grid/src/Grid.tsx [hunk 33]
```diff
@@ -1682,43 +1662,44 @@ class Grid extends PureComponent<GridProps, GridState> {
...
+    window.removeEventListener('mouseup', this.handleMouseUp, true);
 
     if (event.button != null && event.button !== 0) {
       return;
     }
 
-    this.notifyMouseHandlers('onUp', event, false);
+    this.notifyMouseHandlers(
+      'onUp',
+      (event as unknown) as React.MouseEvent,
```
Comment (mattrunyon) [inline packages/grid/src/Grid.tsx:1700]: Make function param `React.MouseEvent?`

File: packages/grid/src/Grid.tsx [hunk 34]
```diff
@@ -1739,7 +1720,7 @@ class Grid extends PureComponent<GridProps, GridState> {
   }
 
   handleWheel(event: WheelEvent): void {
-    this.notifyMouseHandlers('onWheel', event);
+    this.notifyMouseHandlers('onWheel', (event as unknown) as React.WheelEvent);
```
Comment (mattrunyon) [inline packages/grid/src/Grid.tsx:1723]: `React.WheelEvent` param? I think either these should all be react events in the param, or `notifyMouseHandlers` should expect base browser events if they aren't expected to always be react events

File: packages/grid/src/Grid.tsx [hunk 35]
```diff
@@ -1887,9 +1868,13 @@ class Grid extends PureComponent<GridProps, GridState> {
    * @param value New value set
    */
   handleEditCellChange(value: string): void {
-    this.setState(({ editingCell }) => ({
-      editingCell: { ...editingCell, value } as EditingCell,
-    }));
+    this.setState(({ editingCell }) => {
+      assertIsDefined(editingCell);
```
Comment (mattrunyon) [inline packages/grid/src/Grid.tsx:1872]: Can you check that this doesn't break React if `editingCell` is undefined? Or just throw an error in here and see what happens. I have no idea what react does in that case, but I'm assuming it's ok and just reports the error without breaking everything else

File: packages/grid/src/errors/AssertionError.ts [hunk 36]
```diff
@@ -0,0 +1,5 @@
+export class AssertionError extends Error {
+  isAssertionFailed = true;
+}
```
Comment (mattrunyon) [inline packages/grid/src/errors/AssertionError.ts:3]: This is also defined in `assertions.ts`. Consolidate them

File: packages/grid/src/errors/assertIsDefined.ts [hunk 37]
```diff
@@ -0,0 +1,11 @@
+import AssertionError from './AssertionError';
+
+export function assertIsDefined<T>(val: T): asserts val is NonNullable<T> {
+  if (val === undefined || val === null) {
+    throw new AssertionError(
```
Comment (mattrunyon) [inline packages/grid/src/errors/assertIsDefined.ts:5]: This is also in `assertions.ts`. Guessing you wanted to delete `assertions.ts` and keep this + `AssertionError` files?

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts (hunk 28 above)
Comment (mofojed) [reply packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts:72]: I'm not sure I understand this comment

File: packages/grid/src/Grid.tsx (hunk 34 above)
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1723]: Yea some of these handlers are added directly on the window (handleMouseUp) or need to be added directly to get the `passive` option (handleWheel), so they are in fact expecting window events... Really it doesn't matter which kind of event it is, `MouseEvent | React.MouseEvent`, since the properties we check are on either.
I'll make a `type GridMouseEvent = MouseEvent | React.MouseEvent` so it shouldn't matter

File: packages/grid/src/Grid.tsx (hunk 35 above)
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1872]: It breaks everything. I'll add a try/catch around it

File: packages/grid/src/Grid.tsx (hunk 32 above)
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1682]: See my other comment about making a GridMouseEvent

File: packages/grid/src/Grid.tsx (hunk 33 above)
Comment (mofojed) [reply packages/grid/src/Grid.tsx:1700]: Made a GridMouseEvent to handle these cases

File: packages/grid/src/GridUtils.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/grid/src/GridUtils.test.ts:178]: If It will fail the test I'd say that's good enough? Looks like the sort would treat `null` as `0` index effectively

File: packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts (hunk 28 above)
Comment (mattrunyon) [reply packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts]: It doesn't really matter in this specific case since it's a mock for the styleguide, but here's an example for future things.

```ts
await delete([]);
console.log('Unreachable code');
```

If it's `Promise<never>`, then this code doesn't warn you that the 2nd line is unreachable.

If it's just `never`, then VSCode will tell you the 2nd line is unreachable since the 1st function returns never. Never is a subtype of every type, so anything that would be expecting a promise wouldn't complain about types, only unreachable code after it I believe.

Unreachable code isn't currently flagged in TS at least for us b/c ESLint disabled it since TS used to error on compile w/ unreachable code. They since changed that behavior it seems so we should either enable it in tsconfig or enable it in eslint for TS so we avoid having unreachable code

File: packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts [hunk 38]
```diff
@@ -1,6 +1,7 @@
 /* eslint class-methods-use-this: "off" */
 /* eslint no-unused-vars: "off" */
 /* eslint no-empty-function: "off" */
+// eslint-disable-next-line max-classes-per-file
```
Comment (mattrunyon) [inline packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts:4]: Why was this added when nothing else in this file changed?

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts (hunk 38 above)
Comment (mofojed) [reply packages/code-studio/src/styleguide/MockIrisGridTreeModel.ts]: Whoops

File: packages/grid/src/GridUtils.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/grid/src/GridUtils.test.ts:178]: Yea it will fail

**Review by mattrunyon — DISMISSED** [review]

**Comment by mofojed** [comment]
Rebased, just need that +1 again

**Review by spasovski — APPROVED** [review]
